src/estimators/freq_log_log.py

```diff
@@ -0,0 +1,321 @@
+                                                                     'length')
+    assert self.my_hasher == other.my_hasher, "Vectors don't have same hash"
+
+  def get_idx_bits(self, x):
+    # The 'p' largest bits decide which bucket x goes into
+    return (x & self._idx_helper_int) >> self.num_bucket_bits
+
+  def get_w_bits(self, x):
+    # The 'INT_BITS - p' smallest bits decide the value of the bucket
+    return x & self._w_helper_int
+
+  def _count_leading_zeros(self, int_in):
+    most_significant_bit_idx = math.floor(math.log2(int_in))
+    return self.num_bucket_bits - most_significant_bit_idx - 1
+
+  def _compute_rho_value(self, hashed_value):
+    # number of leading zeros + 1 in the bits that come after the p index bits
+    w = self.get_w_bits(hashed_value)
+    if w == 0:
+      # This means all the bits are zero, return total number of bits in bucket
+      return self.num_bucket_bits + 1
+    return self._count_leading_zeros(w) + 1
+
+  def add(self, value):
+    # Handle sparse mode first if needed
+    if self.sparse_mode:
+      self.temp_set[value] = self.temp_set.get(value, 0) + 1
+      # Check if we're at the threshold yet for normal mode
+      if len(self.temp_set) > self.vector_length_m * 6:
+        self.sparse_mode = False
+        self.temp_set.clear()
+
+    # Then do normal HLL calculation no matter if sparse mode is on
+    # (another simplification we made from the paper at the cost of memory)
+    hash_value = self.my_hasher(value)
+
+    bucket_index = self.get_idx_bits(hash_value)
+    bucket_value_part = self._compute_rho_value(hash_value)
+    fingerprint_value = hash_value % 2**32
+
+    # Get the appropriate bucket value:
+    current_value = self.buckets[bucket_index, LEADING_ZEROS_IDX]
+    current_fingerprint = self.buckets[bucket_index, FINGERPRINT_IDX]
+    # if new value is bigger then reset the fingerprint and frequency
+    # or the bucket values are the same the fingerprint is bigger
+    # then reset the fp and the frequency
+    if (bucket_value_part > current_value or
+        bucket_value_part == current_value and fingerprint_value > current_fingerprint):
+      self.buckets[bucket_index, :] = [bucket_value_part, fingerprint_value, 1]
+    # if hashes are the same then increment frequency
+    elif current_fingerprint == fingerprint_value:
+      self.buckets[bucket_index, FREQUENCY_IDX] += 1
+
+  def _raw_hll_cardinality_estimate(self):
+    return self.alpha * self.vector_length_m**2 / np.sum(2.0**(-self.buckets[:, LEADING_ZEROS_IDX]))
+
+  def _linear_counting_estimate(self, arg_1_m, arg_2_v):
+    # Args not always m and V in paper, hence the arg_i_x notation
+    return arg_1_m * np.log(arg_1_m / arg_2_v)
+
+  def _estimate_bias(self, x):
+    _ = self, x
+    # Not defined for now...
+    return 0.0
+
+  def estimate_cardinality_float(self):
+    """Returns k+ reach for k <= 15 as possibly floating point numbers.
```

Comment:
perhaps clarify the returns: a length 15 list where the (k-1)th element is the k+ reach